libqrexec/ioall.c

```diff
@@ -90,6 +98,103 @@ int read_all(int fd, void *buf, int size)
     return 1;
 }
 
+// FIXME: this code is seemingly-correct but needs careful review
+void *qubes_read_all_to_malloc(int fd, size_t initial_buffer_size, size_t max_bytes, size_t *len)
+{
+    size_t buf_size = initial_buffer_size;
+    size_t offset = 0;
+#if PTRDIFF_MAX < INT_MAX
+#error unsupported platform
+#endif
+    if (max_bytes > (size_t)INT_MAX) {
+        LOG(ERROR,
+            "Maximum buffer size %zu exceeds INT_MAX (%d)",
+            max_bytes,
+            INT_MAX);
+        abort();
+    }
+    if (buf_size < 2 || buf_size > max_bytes) {
+        LOG(ERROR,
+            "Minimum buffer size must between 2 and maximum buffer size (%zu) inclusive, got %zu",
+            max_bytes,
+            buf_size);
+        abort();
+    }
+    char *buf = malloc(buf_size);
+    if (buf == NULL) {
+        LOG(ERROR, "malloc() for %zu bytes failed!", buf_size);
+        abort();
+    }
+    *len = 0;
+    for (;;) {
+        size_t to_read = buf_size - offset;
+        ssize_t res = read(fd, buf + offset, to_read);
+        if (res < 0) {
+            // save errno as PERROR() and free() might clobber it
+            int e = errno;
+            if (res != -1)
+                abort(); // kernel bug or some sort of corruption
+            if (e == EINTR || e == EAGAIN || e == EWOULDBLOCK)
+                continue;
+            PERROR("recv");
+            free(buf);
+            errno = e;
+            buf = NULL;
+            break;
+        }
+        size_t const bytes_read_this_time = (size_t)res;
+        if (bytes_read_this_time == 0) {
+            /* EOF */
+            buf[offset] = 0;
+            *len = offset;
+            break;
+        }
+        if (bytes_read_this_time > to_read)
+            abort(); // kernel bug or some sort of corruption
+        if (bytes_read_this_time == to_read) {
+            /* Buffer full.  See if a new buffer can be allocated. */
+            if (buf_size >= max_bytes) {
+                /* Nope, limit reached. */
+                LOG(ERROR, "Too many bytes read (limit %zu)", max_bytes);
```

Comment:
Shouldn't you print `max_bytes-1`? Reading exactly `max_bytes` is already rejected (and rightly so, as you no longer subtract 1 from the `buf_size` for the terminating NUL character).